physicalplan/binary/scalar.go

```diff
@@ -0,0 +1,120 @@
+package binary
+
+import (
+	"context"
+	"sync"
+
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql/parser"
+	"github.com/thanos-community/promql-engine/physicalplan/model"
+)
+
+// scalarOperator evaluates expressions where one operand is a scalarOperator.
+type scalarOperator struct {
+	seriesOnce sync.Once
+	series     []labels.Labels
+	scalar     float64
+
+	pool           *model.VectorPool
+	numberSelector model.VectorOperator
+	next           model.VectorOperator
+	getOperands    getOperandsFunc
+	operation      operation
+}
+
+func NewScalar(pool *model.VectorPool, next model.VectorOperator, numberSelector model.VectorOperator, op parser.ItemType, scalarSideLeft bool) (*scalarOperator, error) {
+	binaryOperation, err := newOperation(op)
+	if err != nil {
+		return nil, err
+	}
+	var getOperands getOperandsFunc
+	if scalarSideLeft {
+		getOperands = getOperandsScalarLeft
+	} else {
+		getOperands = getOperandsScalarRight
+	}
+
+	// Cache the result of the number selector since it
+	// will not change during execution.
+	v, err := numberSelector.Next(context.Background())
+	if err != nil {
+		return nil, err
+	}
+	scalar := v[0].Samples[0]
+
+	return &scalarOperator{
+		pool:           pool,
+		next:           next,
+		scalar:         scalar,
+		numberSelector: numberSelector,
+		operation:      binaryOperation,
+		getOperands:    getOperands,
+	}, nil
+}
+
+func (o *scalarOperator) Series(ctx context.Context) ([]labels.Labels, error) {
+	var err error
+	o.seriesOnce.Do(func() { err = o.loadSeries(ctx) })
+	if err != nil {
+		return nil, err
+	}
+	return o.series, nil
+}
+
+func (o *scalarOperator) Next(ctx context.Context) ([]model.StepVector, error) {
+	in, err := o.next.Next(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if in == nil {
+		return nil, nil
+	}
+	o.seriesOnce.Do(func() { err = o.loadSeries(ctx) })
+	if err != nil {
+		return nil, err
+	}
+
+	out := o.pool.GetVectorBatch()
+	for _, vector := range in {
+		step := o.pool.GetStepVector(vector.T)
```

Comment:
Ideally we should reuse the same vector we get from the input, and modify the values in place.
But for that, we need to have a better way of recycling step vectors so that we know in which pool a vector needs to be returned. 

Maybe the step vector itself can have a `Recycle()` method which will put it back into the correct pool.

Reply:
Yup, this seems like a great idea! This could be iterated on later as well.